src/transformStateWithClones.js

```diff
@@ -8,6 +8,32 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let previousState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(previousState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const remove of action.keysToRemove) {
+          delete previousState[remove];
+        }
+        break;
+
+      case 'clear':
+        previousState = {};
+        break;
+
+      default:
+        throw new Error(`Unhandled action type: ${action.type}`);
+    }
+    history.push({ ...previousState });
```

Comment:
The action of cloning the state copy to the history array is repeated in every case of the switch statement. To adhere to the 'Don't Repeat Yourself' principle, you should perform this action once after the switch statement.